core/container.go

```diff
@@ -0,0 +1,386 @@
+type Container struct {
+	ID ContainerID `json:"id"`
+}
+
+// ContainerID is an opaque value representing a content-addressed container.
+type ContainerID string
+
+// containerIDPayload is the inner content of a ContainerID.
+type containerIDPayload struct {
+	FS     *pb.Definition    `json:"fs"`
+	Config specs.ImageConfig `json:"cfg"`
+	Mounts []ContainerMount  `json:"mounts"`
+}
+
+type ContainerMount struct {
+	Source *pb.Definition `json:"source"`
+	Target string         `json:"target"`
+}
+
+// metaMount is the special path that the shim writes metadata to.
+const metaMount = "/dagger"
+
+func (id ContainerID) decode() (*containerIDPayload, error) {
+	var payload containerIDPayload
+	if err := decodeID(&payload, id); err != nil {
+		return nil, err
+	}
+
+	return &payload, nil
+}
+
+// ContainerAddress is a container image address.
+type ContainerAddress string
+
+func NewContainer(ctx context.Context, st llb.State, cfg specs.ImageConfig, mounts ...ContainerMount) (*Container, error) {
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := encodeID(containerIDPayload{
+		FS:     def.ToPB(),
+		Config: cfg,
+		Mounts: mounts,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Container{
+		ID: ContainerID(id),
+	}, nil
+}
+
+func (container *Container) Decode() (llb.State, specs.ImageConfig, []ContainerMount, error) {
+	if container.ID == "" {
+		return llb.Scratch(), specs.ImageConfig{}, nil, nil
+	}
+
+	payload, err := container.ID.decode()
+	if err != nil {
+		return llb.State{}, specs.ImageConfig{}, nil, err
+	}
+
+	defop, err := llb.NewDefinitionOp(payload.FS)
+	if err != nil {
+		return llb.State{}, specs.ImageConfig{}, nil, err
+	}
+
+	return llb.NewState(defop), payload.Config, payload.Mounts, nil
+}
+
+type containerSchema struct {
+	*baseSchema
+}
+
+var _ router.ExecutableSchema = &containerSchema{}
+
+func (s *containerSchema) Name() string {
+	return "container"
+}
+
+func (s *containerSchema) Schema() string {
+	return schema.Container
+}
+
+func (s *containerSchema) Resolvers() router.Resolvers {
+	return router.Resolvers{
+		"ContainerID":      stringResolver(ContainerID("")),
+		"ContainerAddress": stringResolver(ContainerAddress("")),
+		"Query": router.ObjectResolver{
+			"container": router.ToResolver(s.container),
+		},
+		"Container": router.ObjectResolver{
+			"from":                 router.ToResolver(s.from),
+			"rootfs":               router.ToResolver(s.rootfs),
+			"directory":            router.ErrResolver(ErrNotImplementedYet),
+			"user":                 router.ErrResolver(ErrNotImplementedYet),
+			"withUser":             router.ErrResolver(ErrNotImplementedYet),
+			"workdir":              router.ToResolver(s.workdir),
+			"withWorkdir":          router.ToResolver(s.withWorkdir),
+			"variables":            router.ToResolver(s.variables),
+			"variable":             router.ErrResolver(ErrNotImplementedYet),
+			"withVariable":         router.ToResolver(s.withVariable),
```

Comment:
nitpick: I'm unsure about this name; I think `withEnv` would be much clearer (consistent with Go `os/exec` and container stuff).

"Variable" (without "env" or "environment") feels a little too broad and makes me wonder if there's a template system or something lurking in the darkness. I don't even know what that would mean, but the fact that it doesn't say "env" makes me think something else is going on. :sweat_smile: 

I'm not sure how I'd distinguish between `variable` and `variables` though. And there's no way I'm going to suggest `environmentVariables`.

Some ideas:

1. `withEnv`, `env` (no singular fetch, filter client-side)
2. `withEnv`, `env`, `envVar`
3. `withEnvVar`, `env`, `envVar`

Reply:
Yeah, I agree, I made the same comment on the original PR https://github.com/dagger/cloak/pull/163#issuecomment-1237975539

@shykes Can we change your mind? :)

Reply:
Sure, I was trying to avoid the awkwardness of `env` / `envs` etc.

I don’t love any of the proposed alternatives, but don’t hate them either. Happy to follow the group’s consensus here.

Reply:
I guess `env` is sort of ubiquitous at this point (dockerfiles, the image spec, the bash command `env`, a lot of stdlib methods like nodejs, etc.) so it might be the least surprising. So I guess I'd vote for `#3` above, but it's really just something we need to give to users and get feedback on.

Reply:
> I guess `env` is sort of ubiquitous at this point (dockerfiles, the image spec, the bash command `env`, a lot of stdlib methods like nodejs, etc.) so it might be the least surprising. So I guess I'd vote for `#3` above

What would `env` return? Array of key strings, array of `key=value` strings, array of `EnvVariable` objects with key and value fields?

Reply:
Given the pervasiveness of auto-complete in both editors and playground, I would prioritize clarity and consistency over character count, which leads me to option #4:

```graphql
extend type Container {
  "Return the value of an env variable"
  envVariable(name: String!): String

  "This container plus a new environment variable set"
  withEnvVariable(name: String!, value: String!): Container!

  "A list of all environment variables"
  envVariables: [String!]!
}
```

Reply:
> Given the pervasiveness of auto-complete in both editors and playground, I would prioritize clarity and consistency over character count, which leads me to option 4:

Yeah I'm onboard with that too. I don't personally mind long autocompletable names.

For the return type, I like the idea that the format you set an env var with is the same as the format we return it to you in. So if we return `["k=v"]` then maybe that means we should accept the input as `withEnvVariable(var: "k=v")` too? Or that we should return an array of `EnvVariable` objects and keep `withEnvVariable` as is.

This is really *really* splitting hairs at this point though :-)

Reply:
I'm good with spelling it out, it was just the missing "env" that threw me for a loop. :)

> So if we return ["k=v"] then maybe that means we should accept the input as withEnvVariable(var: "k=v") too? Or that we should return an array of EnvVariable objects and keep withEnvVariable as is.

I prefer the higher-level `withEnvVariable(name: String!, value: String!): Container!` since this does a little more than just appending a `k=v` to a list: it actually replaces the named var if it's present. So I think it's OK for it to have a higher-level representation.

It'll also be easier to pass a parameterized value in, since I'm not sure how to do string concatenation in GraphQL.

I think there's a decent argument for having `envVariables` return a list of key-value pairs, just to be more symmetrical, but I'm not too worried about it.

---

I've split this out to https://github.com/dagger/dagger/issues/3204 for now since I'd like to merge this PR so I can push today's changes.